platform-sdk/docs/proposals/TSS-Block-Signing/tss-block-signing.md

```diff
@@ -0,0 +1,286 @@
+- Upon availability of the implementation:
+  - This proposal's full behavior can be tested in integration tests.
+
+Impacts to the Services Team:
+
+- The ability to sign blocks with TSS is blocked until this proposal delivers its capability.
+
+Implication Of Completion:
+
+- Signing blocks and construction of block proofs with TSS signatures is unblocked.
+
+### Requirements
+
+The core requirements governing the TSS proposals are stated in `TSS-Ledger-ID`.
+
+Blocks Signing Specific Requirements:
+
+1. A block is considered signed if its root hash has a ledger signature or a following block's root hash has a
+   ledger signature.
+2. TSS ledger signatures are best-effort with low probability of failure.
+3. The block proofs produced by different nodes for the same block do not need to be identical.
+4. Nodes attempt to sign every block.
+5. The number of consecutive blocks that are not signed is minimized.
+
+### Design Decisions
+
+No Guaranteed Ledger Signature:
+
+- `TssShareSignatureTransaction` are handled pre-consensus in the pre-handle phase of transaction handling
+  - this is to allow the creation of signatures as fast as possible.
+- The `TssShareSignatureTransaction` are not stored in the state.
+
+#### Alternatives Considered
+
+Guaranteed Ledger Signature Generation:
+
+- requires storing share signatures in the state.
+- requires processing share signatures after consensus, introducing a delay in block signing.
+
+This was not selected for the following reasons:
+
+1. block proofs do not need to be identical between nodes.
+2. the probability of failure to sign a block is low.
+3. speed in signing a block matters since blocks are not published without signatures once they are signable.
+
+## Changes
+
+The changes are presented in the following order:
+
+1. Public API
+2. Component Architecture
+3. Core Behaviors
+4. Configuration
+5. Metrics
+6. Performance
+
+### Public API
+
+#### Services
+
+The `TssBaseService` is extended with the following API:
+
+```Java
+/**
+ * The TssBaseService will attempt to generate TSS key material for any set candidate roster, giving it a ledger id and
+ * the ability to generate ledger signatures that can be verified by the ledger id.  Once the candidate roster has
+ * received its full TSS key material, it can be made available for adoption by the platform.
+ * </p>
+ * The TssBaseService will also attempt to generate ledger signatures by aggregating share signatures produced by
+ * calling {@link #requestLedgerSignature(byte[])}.
+ *
+ */
+public interface TssBaseService {
+
+    ...
+
+    /**
+     * Requests a ledger signature on a message hash.  The ledger signature is computed asynchronously and returned
+     * to all consumers that have been registered through {@link #registerLedgerSignatureConsumer(Consumer)}.
+     *
+     * @param messageHash The hash of the message to be signed by the ledger.
+     */
+    void requestLedgerSignature(@Nonnull byte[] messageHash);
```

Comment:
Minor: why not `Bytes`?

Reply:
Fair. However, I'm not sure if this particular API is going to be invoked so very frequently that an extra object allocation would present a performance issue.

The reason why `Bytes` might be preferred is because it represents an **immutable** array of bytes as opposed to a plain `byte[]` object.

Also, I'm unsure if an underlying low-level library requirements should define the high-level API input to this extent.

Having said that, I don't know which objects the caller handles. If they start with `byte[]`, then indeed, this might be the way to go. However, if they have a `Bytes` and then have to allocate a new bytes array just to pass it to this API, then perhaps we should reconsider.

Reply:
Even if we input in `Bytes`, the underlying library API is in `byte[]`, so the conversion has to happen anyway.